src/coreclr/vm/method.cpp

```diff
@@ -2485,526 +2636,213 @@ MethodDesc* MethodDesc::GetMethodDescFromStubAddr(PCODE addr, BOOL fSpeculative
-
-            pCode->Setup(pCodeRX, pBaseMD, GetPreStubEntryPoint(), pLoaderAllocator);
-
-            p += TEP_CENTRAL_JUMP_SIZE;
-
-            relOffset -= TEP_CENTRAL_JUMP_SIZE;
+            InterlockedCompareExchangeT(pSlot, tempEntryPoint, (PCODE)NULL);
         }
-
-        indexInBlock++;
-
-#endif // TARGET_ARM
-
-        relOffset -= TEP_ENTRY_SIZE;
-        pMD = (MethodDesc *)((BYTE *)pMD + pMD->SizeOf());
+        InterlockedUpdateFlags4(enum_flag4_TemporaryEntryPointAssigned, TRUE);
     }
-
-#ifdef TARGET_ARM
-
-    CentralJumpCode* pCode = (CentralJumpCode*)p;
-    memcpy(pCode, &c_CentralJumpCode, TEP_CENTRAL_JUMP_SIZE);
-    pCode->Setup (GetPreStubCompactARMEntryPoint(), this);
-
-    _ASSERTE(p + TEP_CENTRAL_JUMP_SIZE == (BYTE*)temporaryEntryPointsWriterHolder.GetRW() + size);
-
-#else // TARGET_ARM
-
-    _ASSERTE(p == (BYTE*)temporaryEntryPointsWriterHolder.GetRW() + size);
-
-#endif // TARGET_ARM
-
-    ClrFlushInstructionCache((LPVOID)temporaryEntryPoints, size);
-
-    SetHasCompactEntryPoints();
-    return temporaryEntryPoints;
 }
-#endif // !DACCESS_COMPILE
-
-#endif // HAS_COMPACT_ENTRYPOINTS
 
 //*******************************************************************************
-PCODE MethodDescChunk::GetTemporaryEntryPoint(int index)
+void MethodDescChunk::DetermineAndSetIsEligibleForTieredCompilation()
 {
-    LIMITED_METHOD_CONTRACT;
+    WRAPPER_NO_CONTRACT;
 
-#ifdef HAS_COMPACT_ENTRYPOINTS
-    if (HasCompactEntryPoints())
+    if (!DeterminedIfMethodsAreEligibleForTieredCompilation())
     {
-#ifdef TARGET_ARM
+        int count = GetCount();
 
-        return GetTemporaryEntryPoints() + COMPACT_ENTRY_ARM_CODE + THUMB_CODE + index * TEP_ENTRY_SIZE;
-
-#else // TARGET_ARM
-
-        int fullBlocks = index / TEP_MAX_BLOCK_INDEX;
-        int remainder = index % TEP_MAX_BLOCK_INDEX;
-
-        return GetTemporaryEntryPoints() + 1 + (fullBlocks * TEP_FULL_BLOCK_SIZE) +
-            (remainder * TEP_ENTRY_SIZE) + ((remainder >= TEP_MAX_BEFORE_INDEX) ? TEP_CENTRAL_JUMP_SIZE : 0);
-
-#endif // TARGET_ARM
-    }
-#endif // HAS_COMPACT_ENTRYPOINTS
+        // Determine eligibility for tiered compilation
+        {
+            MethodDesc *pMD = GetFirstMethodDesc();
+            bool chunkContainsEligibleMethods = pMD->DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk();
 
-    return Precode::GetPrecodeForTemporaryEntryPoint(GetTemporaryEntryPoints(), index)->GetEntryPoint();
-}
+    #ifdef _DEBUG
+            // Validate every MethodDesc has the same result for DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk
+            MethodDesc *pMDDebug = GetFirstMethodDesc();
+            for (int i = 0; i < count; ++i)
+            {
+                _ASSERTE(chunkContainsEligibleMethods == pMDDebug->DetermineIsEligibleForTieredCompilationInvariantForAllMethodsInChunk());
+                pMDDebug = (MethodDesc *)(dac_cast<TADDR>(pMDDebug) + pMDDebug->SizeOf());
```

Comment:
Is the `dac_cast<>` call appropriate here? This doesn't use `PTR_MethodDesc`, whcih means the call to `SizeOf()` isn't correct. It seems this should really be a `static_cast<>`.

Reply:
static_cast can be used here instead. This code isn't compiled for the dac, so its all the same at runtime.

Reply:
We can't actually use static_cast here, so I can use either a C style cast, reinterpret_cast, or dac_cast. Given that we commonly have logic which casts to TADDR which is required to use dac_cast, I'd like to keep the existing code, so it will work if anyone copies/pastes this code into a place where the dac actually needs to use it.

Reply:
Okay. Let's keep it.